### ISSUE TITLE
EEX (Even Exchanges)

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1039]
_Imported from trac ticket 1039.
Created by **rneville**
Opened in trac: Aug 30, 2011, 11:08:11 AM
Last modified in trac: Nov 21, 2013, 2:49:53 PM_
***
_by rneville:_
See [design](comment:16).

***
* kbower attached [eex.lst](https://trac.retailarchitects.com/trac/attachment/ticket/1039/eex.lst) on Nov 12, 2013, 4:01:36 PM

**Comment by kbower** [comment]
_**rneville** trac comment on Aug 30, 2011, 11:25:08 AM:_

 * item swaps, no artrans _(Edit: other than EEX document's artran)_, schedule implies a pickup and a delivery date simultaneously

**Comment by kbower** [comment]
_**rneville** trac comment on Aug 30, 2011, 11:46:45 AM:_

 * needs an incoming/outgoing location upon finalization

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 2:36:59 PM:_

Hi JR,
Below is a list of *current* (not archived) AR_TRN records.

I'm trying to understand EEX.  These are marked as EEX, but have no SO document that match.  Can you explain these?

Thanks!
Kent

```
ORDERID        ADJINVOICE     TRA
-------------- -------------- ---
06053030007    0617303DLNR    EEX
10242060047    1107206ATSZ    EEX
07304000105    0803400JTUD    EEX
06053030006    0804303EBTS    EEX
06302080006    0707208ACIH    EEX
0830003FXHL    0907003GGAMA   EEX
07223040017    0728304DZSO    EEX
06053030007    0614301DKSD    EEX
04063030046    0615303DKXK    EEX
07223040017    0728304DZRV    EEX
10172030157    1029203ARNI    EEX
10182040054A   1102204ASJH    EEX
10172030156    1104203ATBQ    EEX
06053030007    0610303DJLX    EEX
06053030006    0610303DJMB    EEX
04063030046    0609303DIPB    EEX
04063030048    0609303DIPT    EEX
10182040054A   1116204AVJJ    EEX
12202060044    0708301DSQJ    EEX
06053030007    0610301DJOJ    EEX
06053030007    0610301DJOR    EEX
06053030006    0610301DJOY    EEX
06283060027    0727306DZQV    EEX
10223030012    0112403GNDH    EEX
06053030006    0618303DLUZ    EEX
06053030006    0610301DJOG    EEX
0525300DCBO    0531300DEQR    EEX
10182040054A   0224304BWGD    EEX
06283060027    0804300EBST    EEX
11013030028    1110303FJAR    EEX
11013030028    1106303FHVR    EEX
10313030002    1103303FGLO    EEX
10223030012    1201303FTBX    EEX

33 rows selected.
```

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 2:39:03 PM:_

Hi Wayne,
Below is a list of *current* (not archived) AR_TRN records.

I'm trying to understand EEX.  These are marked as EEX, but have no SO document that match.  Can you explain these?

Thanks!
Kent

```
ORDERID        ADJINVOICE     TRA
-------------- -------------- ---
09032140021    10072140021    EEX
02165050006    0616505MAUV    EEX
10134150053    02275150052    EEX
04304150189    0629515MTGY    EEX
05045150044    0621515MICI    EEX
05162020018    06022020039    EEX
10172140012    11032140012    EEX
08302140009B   02053140016    EEX
11094140076    01185140059    EEX
04093050100    06103050024    EEX
04093050098    06103050025    EEX
12044150033    01145150061    EEX
01115060036    01205060043    EEX
12274150048    05135150002    EEX
11194150062    06085150025    EEX
01075100024    01245107007    EEX
               01225150024    EEX
10014150065    11114150017    EEX
0723506OBPKC   0901506QHRS    EEX
12274150048    05125150023    EEX
12274150048    05105150049    EEX
05045150044    05275150104    EEX
12274150048    03145150099    EEX
0723506OBPKC   0916506RDXV    EEX
04304150189    0621515MIGT    EEX
05045150044    0630515MUSB    EEX
1104506TZXC    1226506XCKC    EEX
0716501NQSW    0416601EGFX    EEX
1125503VHTL    0816603KIIM    EEX
1124510VEYK    1120610PCJO    EEX
240469189      0530712YIRE    EEX
1209002PJEF    0214102YEFW    EEX

32 rows selected.
```

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 2:44:06 PM:_

Note that in 63 of 65 cases listed above, no matching SO or SO_ARC record (of any type) was found for either orderid or adjinvoice:

```
select a.ORDERID, a.ADJINVOICE, TRANSACTIONCODEID from
artrans a
WHERE TRANSACTIONCODEID = 'EEX'
and not exists (select 0 from orders b where a.ADJINVOICE = b.orderid )
and not exists (select 0 from so_arc b where a.ADJINVOICE = b.del_doc_num )
and not exists (select 0 from orders b where a.ORDERID = b.orderid )
and not exists (select 0 from so_arc b where a.ORDERID = b.del_doc_num )
```
It seems that ADJINVOICE is the correct field for the orderid of the EEX, pending explanation of the above.

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 2:48:26 PM:_

Find cases where the orderid matches an EEX, and the ADJINVOICE does NOT:
```
select a.ORDERID, a.ADJINVOICE, TRANSACTIONCODEID from
artrans a
WHERE TRANSACTIONCODEID = 'EEX'
and not exists (select 0 from orders b where a.ADJINVOICE = b.orderid and b.type = 'EEX')
and exists (select 0 from orders b where a.ORDERID = b.orderid and b.type = 'EEX')
```
Seems to consistently return no records for any retailer.

Again, support that ADJINVOICE is the correct field for the orderid of the EEX

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 3:02:04 PM:_

Replying to [kbower](comment:7):
> Note that in 63 of 65 cases listed above, no matching SO or SO_ARC record (of any type) was found for either orderid or adjinvoice:
> 

In the strange case for kanes (1209002PJEF    0214102YEFW    EEX),
the orderid exists as an SO_ARC SAL.  So still, this does not disqualify our assumption that ADJINVOICE is the correct field for the orderid of the EEX.

In the strange case for eld (0830003FXHL    0907003GGAMA   EEX), both 0830003FXHL     and 0907003GGAMA are SALEs.  ! ? Maybe this was someone hacking with SQL??

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 3:24:17 PM:_

If you look at the DATES of these documents they are from 15 to 20 years ago.

The EEX records were probably never Archived for some reason, and the SO record was PURGED.

(We used to PURGE SO records every few years)

It's not that we enter the  AR_TRN record for EEX separately from the SO.

Does this answer?

Jesús R. Capó, VP, CIO

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 3:35:11 PM:_

Thanks, that is very helpful.  I had assumed there would be no AR_TRN records that old in AR_TRN...that they would have been archived, so I didn't look at the dates.

The only one that still is a mystery is this one:
```
IVC_CD         ADJ_IVC_CD     TRA POSTDATE
-------------- -------------- --- --------------------
0830003FXHL    0907003GGAMA   EEX 07-SEP-2000 00:00:00
```
Both 0830003FXHL and 0907003GGAMA are 'SAL' SO documents.  So what is this EEX doing in AR_TRN??

Maybe you can answer my **real question:
When an EEX is written, its document number goes in the corresponding AR_TRN's ADJ_IVC_CD field OR the IVC_CD field?**  Which one?

Thanks!
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 4, 2012, 4:59:58 PM:_

The EEX number goes in the ADJ_IVC_CD field. (Just like CRM, MCR, MDB)

Jesús R. Capó, VP, CIO

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 5, 2012, 7:51:12 AM:_

Ticket [#1874](1874) has been marked as a duplicate of this ticket.

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 5, 2012, 7:52:57 AM:_

(Changeset [[3112]](../commit/3ee7d7c709ecf2a1ccf6db65d0dff49c35de086e))
 * fix [#1916](1916) (NoResultFound: No row was found for one() when attempting to open ticket with archived artrans)
 * EEX uses the adjinvoice field in artrans like most other adjustments, see #1039

**Comment by kbower** [comment]
_**kbower** trac comment on Apr 18, 2013, 10:40:11 AM:_

# Design Checklist
 * ~~Break order screen should filter out non inventory lines.~~
 * Delivery calendar implementation
   * ~~must subtract from capacity~~
   * ~~ `kan.py`/rules need to avoid referencing customerorder~~

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2013, 11:34:32 AM:_

outstanding questions to two retailers:
----
I have a few EEX questions:

    Do EEX's reserve in the exact same way as SAL's? 
    At finalization, do the lines create inventory in the same way as CRM's?
    Does the EEX's line get changed from the outgoing piece's LOC and STORE_CD to the incoming pieces?

        Is the original location of the outgoing piece stored anywhere other than transactions or is it lost from the line?

Thanks,
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2013, 12:47:48 PM:_

Kent,

1)      On entry(EEXI): You have to select the location or type it in. On change(EEXC): You can autoreserve which is the same as SAL’S. (I think that entry(EEXI) is a flaw in GERS. You should be able to autoreserve & use SPFM logic.)

2)      Yes. But it does not show the incoming location in the SO_LN table like a CRM does.

3)      No. It shows the outgoing piece location but not the incoming location(SO_LN). However, it does show the incoming in the ITM_TRN table as a ITT Tp of CMF.

Michelle
----
I cant think of a reason why there should be a spiff line. It may just be defaulting from the SO_LN logic but since it is an exchange it shouldn’t matter.

 
Also, I have a correction on #1 below [above]. Having to select the location. That was a system parameter question to show Location only or both Store & Location. When I changed it, the autoreserve feature works now. Sorry. So to answer that question, YES. EEX reserves the same as SAL’S.

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2013, 2:24:14 PM:_

If anyone has input, now is the time (or we will continue with our own assumptions):

We are implementing Exchanges and the idea of a spiff on an exchange line doesn't seem relevant to us, so we may just leave it unconditionally blank.  I've noted that GERS seems to fill it in, but we don't know what that means.

If there is meaning to an Exchange's spiff, please explain it or else we will just leave it blank.

Thanks!
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2013, 4:30:39 PM:_

As long as it's a pure Even Exchange (same piece being shipped and received - and both are finalized), the SPIFF would have zero effect.

What complicates it would be if you only partially complete the Even Exchange.  For example, if a customer returns BOTH pieces (the CRM portion and the SAL portion) you would have a partially complete Even Exchange.  This would have the same effect as a Credit Memo, and a Voided Sales Order.  In this case, the SPIFF would be a factor.


Jesús R. Capó, VP, CIO
El Dorado a different kind of furniture store.
----
You are correct Kent.  The option to create two documents (SAL and CRM) exists at the point of entry of EEX and in those cases it's not an actual EEX document.  So the point is moot.   EEX do not affect SPIFFS.

To expand a bit, if a customer returns a piece with a driver, we could create an "even exchange" in order to receive today the SKU being returned (CRM), and to have another document ship out a new one the next day (SAL).  GERS currently allows the keying into ONE program (also called even exchange) that creates both of these documents simultaneously  (without having to create two separate orders).  BUT, this does not count as an EEX document.  A standard EEX document is used to pick up and deliver at the same time, and if either of those is not done, then the EEX is Voided or Finalized (they essentially have the same effect) and the corresponding correct document is created.

Sorry for the confusion.

Jesús R. Capó, VP, CIO

**Comment by kbower** [comment]
_**sjohnson** trac comment on Oct 24, 2013, 2:17:13 PM:_

Scott,
It’s been several years since we last used EEX orders but I can tell you that every EEX order we have in the live tables or the archive tables have a NULL value or 0 value in the SO_LN spiff column. We do have spiffs on CRM’s. These would be negative spiffs against the salesperson.

Gene Frazier
IT Department
Levin Furniture
724-872-2050 x234

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 28, 2013, 9:20:20 AM:_

Sal,
I need help with some GER detective work for EEX.

On TRAIN, please:
1. Find a SKU that has four different costs (or create this circumstance).
2. Write a SAL with 2 lines for the SKU, each QTY 1 (different costs), and finalize it.
3. Write an EEX of QTY 1 for the SKU such that it is filled with a 3rd (unique) cost.  (Stop here and let me inspect.)
4. Finalize the EEX. (Stop here and let me inspect.)  (We need to figure out which cost it returns.)
5. Now write another EEX (fill with 4th cost) and finalize that piece.  Is GERS smart enough to return the cost from among the two outstanding costs, or does it incorrectly return the same cost that it did with the first EEX?

Thanks very much!
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Nov 12, 2013, 1:53:46 PM:_

Hmmm...  The 2 that left the building originally cost 510 and 610, but GERS brought the $710 exchange one back... I guess that isn't horrible from accounting, but isn't true to what happened and I think technically incorrect.

Ok, please do step 5 now (make sure the new EEX is filled with the $410 cost piece (not the $710 again).

Thanks.

**Comment by kbower** [comment]
_**kbower** trac comment on Nov 13, 2013, 5:16:36 PM:_

Is it true that finalizing a CRM writes *all three* of these:
UNR
OOC
CMF

But finalizing an EEX *only *writes:
CMF


?

Please let me know,
Thanks!
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Nov 14, 2013, 4:30:20 PM:_

Finalizing in GERS:

EEX creates a CMF and SOF.
CRM creates a UNR and CMF.

I don’t understand the reasoning behind the UNR.  UNR is “unreserved” and it also appears when removing inventory from a Sales Order.

Jesús R. Capó, VP, CIO

**Comment by kbower** [comment]
_**kbower** trac comment on Nov 21, 2013, 2:49:53 PM:_

(Changeset [[4800]](../commit/26d35483b3c5b4da13e5307cffb3e92bf8f3828b))
* closes #1039, (Even) Exchanges #all

**Comment by kbower** [comment]
_**kbower** trac status change on Nov 21, 2013, 2:49:53 PM:_ **closed**
